aptos-move/framework/aptos-framework/sources/object.move

```diff
@@ -0,0 +1,433 @@
+        let current_address = destination;
+        assert!(
+            exists<Object>(current_address),
+            error::not_found(EOBJECT_DOES_NOT_EXIST),
+        );
+
+        while (owner != current_address) {
+            // At this point, the first object exists and so the more likely case is that the
+            // object's owner is not an object. So we return a more sensible error.
+            assert!(
+                exists<Object>(current_address),
+                error::permission_denied(ENOT_OBJECT_OWNER),
+            );
+            let object = borrow_global<Object>(current_address);
+            assert!(
+                object.allow_ungated_transfer,
+                error::permission_denied(ENO_UNGATED_TRANSFERS),
+            );
+
+            current_address = object.owner;
+        };
+    }
+
+    /// Accessors
+
+    public fun owner(object_id: ObjectId): address acquires Object {
+        assert!(
+            exists<Object>(object_id.inner),
+            error::not_found(EOBJECT_DOES_NOT_EXIST),
+        );
+        borrow_global<Object>(object_id.inner).owner
+    }
+
+    public fun is_owner(object_id: ObjectId, owner: address): bool acquires Object {
+        owner(object_id) == owner
+    }
+
+    native fun create_signer(addr: address): signer;
+
+    #[test_only]
+    use std::option::{Self, Option};
+
+    #[test_only]
+    const EHERO_DOES_NOT_EXIST: u64 = 0x100;
+    #[test_only]
+    const EWEAPON_DOES_NOT_EXIST: u64 = 0x101;
+
+    #[test_only]
+    struct HeroEquipEvent has drop, store {
+        weapon_id: Option<ObjectId>,
+    }
+
+    #[test_only]
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct Hero has key {
+        equip_events: event::EventHandle<HeroEquipEvent>,
+        weapon: Option<ObjectId>,
+    }
+
+    #[test_only]
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct Weapon has key { }
+
+    #[test_only]
+    public fun create_hero(creator: &signer): ObjectId acquires Object {
+        let hero_creator_ref = create_named_object(creator, b"hero");
+        let hero_signer = generate_signer(&hero_creator_ref);
+        let guid_for_equip_events = create_guid(&hero_signer);
+        move_to(
+            &hero_signer,
+            Hero {
+                weapon: option::none(),
+                equip_events: event::new_event_handle(guid_for_equip_events),
+            },
+        );
+
+        object_id_from_creator_ref(&hero_creator_ref)
+    }
+
+    #[test_only]
+    public fun create_weapon(creator: &signer): ObjectId {
+        let weapon_creator_ref = create_named_object(creator, b"weapon");
+        let weapon_signer = generate_signer(&weapon_creator_ref);
+        move_to(&weapon_signer, Weapon { });
+        object_id_from_creator_ref(&weapon_creator_ref)
+    }
+
+    #[test_only]
+    public fun hero_equip(
+        owner: &signer,
+        hero: ObjectId,
+        weapon: ObjectId,
```

Comment:
how are you verifying that whatever thing you're equipping here to the hero is actually a weapon?